### ISSUE TITLE
Fatal relocation not allowed

### ISSUE THREAD

**Issue by connorourke** [apptainer/apptainer#1144]
### Version of Apptainer

v1.1.6

### Expected behavior

When I run 
`./apptainer exec docker://alpine cat /etc/alpine-release
`
I expect to see:

``` 
INFO:    Using cached SIF image
3.17.2
```
### Actual behavior

What actually happened? Why was it incorrect?
I get this instead:

```
FATAL:   Relocation not allowed with starter-suid
```

### What OS/distro are you running

Centos 7

```sh
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

```

### How did you install Apptainer

Installed from source:

```
./mconfig --with-suid --prefix=/apps/apptainer
make
sudo make install
```

**Comment by GodloveD** [comment]
is this a regression to #932?

**Comment by connorourke** [comment]
Doesn't look to be - the fix for #934 looks to be in place.

**Comment by DrDaveD** [comment]
You don't show "cd builddir", I assume that's an oversight.

Are you attempting to run apptainer from the builddir?  That won't work.  Does it work if you use `/apps/apptainer/bin/apptainer` or put `/apps/apptainer/bin` in your PATH?

When I run `./apptainer` from the builddir however I get a different error about not being able to open apptainer.conf.  Oh yes that's because there's no apptainer-suid in the build area.  So that's not an explanation for the problem.  I haven't been able to reproduce your problem.  I get no error if I cd to /apps/apptainer/bin and run `./apptainer`.

**Comment by DrDaveD** [comment]
Then what is the "./" for in the apptainer command?

The suid binary is called `starter-suid` and it should be in `/apps/apptainer/libexec/apptainer/bin`.

**Comment by connorourke** [comment]
I'm not running from the install directory, but from the user home, with the bin dir in my PATH. If i run from the install dir i get the same error. The starter-suid binary is there in `/apps/apptainer/libexec/apptainer/bin`.

Does the error refer to having moved the install directory?

**Comment by gdonval** [comment]
Are `sylog.Debugf`/`sylog.Warningf` supposed to output something on screen in the default case? If not, what is necessary to achieve that?

If so, that suggests that `os.Stat` does return a non-`nil` error, which I suspect is related to the peculiarities of `/apps` being a mount point. 

@connorourke, would it be possible to try adding a print/sylog statement [there](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L104) and see what you get?

@DrDaveD, the [code](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L105) as it is returns an empty string AFAIK. Did you perhaps mean to use a named returned value [there](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L91)?

```
func getPrefix() (installPrefix string) {
```

**Comment by connorourke** [comment]
Looks like if I build on the compute node I get this error, but not if I build on the login node. 

I now have a new error though:

``` 
apptainer  exec docker://alpine cat /etc/alpine-release
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Getting image source signatures
Copying blob 63b65145d645 skipped: already exists  
Copying config 6a2bcc1c7b done  
Writing manifest to image destination
Storing signatures
2023/03/16 16:27:22  info unpack layer: sha256:63b65145d645c1250c391b2d16ebe53b3747c295ca8ba2fcb6b0cf064a4dc21c
INFO:    Creating SIF file...
FATAL:   Unable to handle docker://alpine uri: while building SIF from layers: while creating squashfs: create command failed: signal: segmentation fault: 
``` 

Which looks to be an issue with squashfs, not apptainer.

**Comment by gdonval** [comment]
> Oh, you're right, that's a bug.

Actually, turns out it is not because we are in a closure as in `prefixOnce.Do(func(){...})`... So the returns are with respect to the closure with the relevant variable (`installPrefix`) correctly set...

I really struggle to see how this is even possible, [this](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L177) has to be false. 

However the *only* way `getPrefix()` can get called without outputting anything (with `-d`) is through a `os.Stat` failure. That failure will set the return value to `{{.Prefix}}` (I've checked with latest Go, it does work like that). So `prefix == "{{.Prefix}}"` **cannot** be false.

> It was added in version 1.1.6, in https://github.com/apptainer/apptainer/issues/1061.

Ah I see. It's tricky and a proper solution would probably require quite a bit of refactoring.


> Looks like if I build on the compute node I get this error, but not if I build on the login node.

Oh... Could it be that the debug flag is broken then? That would mean `getPrefix` would go through to the end of the closure (without ever outputting the prefix). That would change everything then, all it would take is [this](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L124) becoming false. This is easy: `os.Executable` [is not guaranteed to be stable](https://pkg.go.dev/os#Executable) (it's OS dependent... actually even OS version dependent).

----

All this seems a bit brittle. Why not separating the configuration depending on whether the executables have been compiled with `--with-suid` or not? When it's the case, just use the compile-time prefix. When it's not, use the full implementation you have now. Would that sound alright?

**Comment by DrDaveD** [comment]
If you click on `Executable` in the [doc](https://pkg.go.dev/os#Executable) it brings you to the source code, then click on `executable` to get the real function.  On linux it simply reads from `/proc/self/exe` (possibly stripping off an ending `" (deleted)"`) and reading about that in [proc(5)](https://man7.org/linux/man-pages/man5/proc.5.html) does not indicate any instability.  I think the go doc means it might be different on linux & netbsd, which is not much of a problem.

**Comment by gdonval** [comment]
>  I do however see the "Install prefix is" debug message during a normal, working run.

I have access (as a simple user, unfortunately) to a cluster exhibiting the same behaviour (hence my interest). I can guarantee that the only thing that gets displayed when typing `apptainer -d` was the error reported here.

> I don't follow.

Sorry, I realise I really wasn't clear. What I meant is that there are 3 return points in the `getPrefix` closure. 2 of them trigger the display of some information on screen ([missing exec](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L95) and [end](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L134)). As there is no information displayed, [only one return point](https://github.com/apptainer/apptainer/blob/e6efa4752250e8f5ab607c281fd714a4901709e9/internal/pkg/buildcfg/confgen/gen.go#L105) is possible. 

> I don't understand the implications of that not being stable.

```
$ cd /tmp
$ ln -s /usr/bin localbin
$ ls -la /proc/self/exe
lrwxrwxrwx 1 user group 0 Mar 17 11:56 /proc/self/exe -> /usr/bin/ls
$ localbin/ls -la /proc/self/exe
lrwxrwxrwx 1 user group 0 Mar 17 11:57 /proc/self/exe -> /usr/bin/ls
```

So translating to our case, if `apptainer` (or `/apps` or `/apps/apptainer`, etc.) is symlinked to another place (e.g. `/apps -> /mnt/nfsapps`), calling `/apps/apptainer/bin/apptainer` will make `os.Executable()` return `/mnt/nfsapps`, not `/apps/apptainer`. Then `starter-suid` will complain loudly.

----

----

> The problem with that is...

I am not sure I understand what the original problem and use case is...

Is there any documentation explaining how `apptainer` and `start`/`start-suid` are working together?

**Comment by DrDaveD** [comment]
Ok, I have reproduced the issue by introducing a symlink.  I see what you mean now about no debug being printed with the `-d` option.  That's because the `getPrefix()` function does get called once very early (from `isSuidInstall()`) before printing is initialized, but it also gets called later and that's why I see debug messages when it succeeds.

So I am trying out a fix that resolves the compiled-in paths so they will matched the resolved path of the executable, should be ready soon.

> So I follow until the separation of privilege being implemented as `apptainer` doing all user-level stuff and `start[-suid]` doing all the privileged stuff. That is fine and good design.
> 
> Then I'm completely lost.
> 
> In rootless/relocatable mode, what prevents you from just finding a version-matching `start` binary in the default compile-time prefix, overridable with a flag and/or an environment variable? No complex relocation necessary it seems (keep in mind there are probably aspects I did not consider but that how it does look to me).
> 
> In SUID/fixed mode, what prevents you from just using the compile-time prefix as only source of truth? Would using a version-matching yet different `apptainer` binary pose any security problem?

As I tried to explain above, rpm and deb packages have to be compiled in SUID mode, but install-unprivileged.sh wants to relocate all those same binaries except starter-suid, that is, all the files that are in the `apptainer` rpm but not the one in the `apptainer-suid` rpm.  So it needs a run-time check to make sure that there is no `starter-suid` binary because if it is there, relocation is not allowed.

**Comment by DrDaveD** [comment]
The fix is now available in #1211
